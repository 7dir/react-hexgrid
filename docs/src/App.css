@font-face {
    font-family: 'SauceCodePro Nerd Font';
    src: local("./SauceCodeProNerdFontMono-Black.ttf");
    format: ('truetype');
}

* {
    font-family: "SauceCodePro Nerd Font", serif;
}