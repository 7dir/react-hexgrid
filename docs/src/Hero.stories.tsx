import React from "react";
import type { Meta, StoryObj } from "@storybook/react";
import Hero from "./Hero"

const meta = {
    title: "Stories/Hero",
} satisfies Meta<typeof Hero>

export default meta;

type Story = StoryObj<typeof Hero>

export const Default: Story = {
    render: () => (
        <Hero />
    )
}

