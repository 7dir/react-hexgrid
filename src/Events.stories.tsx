import React from "react"
import type { Meta, StoryFn } from "@storybook/react";
import { HexGrid, Layout, Hexagon, GridGenerator, Hex } from "./";

const meta = {
    title: "Stories/Events",
    component: Hexagon,
} satisfies Meta<typeof Hexagon>;

export default meta;

type Story = StoryFn<typeof Hexagon>;

