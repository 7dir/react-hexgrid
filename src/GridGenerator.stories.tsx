import React from "react";
import type { Meta, StoryObj } from "@storybook/react";

import Hexagon from "./Hexagon/Hexagon";
import { HexGrid, GridGenerator, Layout } from "./"

const meta = {
    title: "Stories/GridGenerator",
} satisfies Meta<typeof GridGenerator>

export default meta;

type Story = StoryObj<typeof GridGenerator>

export const Default: Story = {
    render: () => (
        <>
            <HexGrid width="100%" height="100%">
                <Layout>
                    {GridGenerator.hexagon(2).map((hex, i) => (
                        <Hexagon key={i} q={hex.q} r={hex.r} s={hex.s} cellStyle={{ fill: "red", stroke: "black", strokeWidth: "0.1" }} />
                    ))}
                </Layout>
            </HexGrid>
        </>
    )
}

