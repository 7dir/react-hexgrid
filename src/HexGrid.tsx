import React, { Component } from "react"
import PropTypes from "prop-types"

export type HexGridProps = {
  width?: string | number
  height?: string | number
  viewBox?: string
  children: React.ReactElement | React.ReactElement[]
}
export function HexGrid({
  width = 800,
  height = 600,
  viewBox = "-50 -50 100 100",
  children,
}: HexGridProps) {
  // static propTypes = {
  //   width: PropTypes.oneOfType([
  //     PropTypes.string.isRequired,
  //     PropTypes.number.isRequired,
  //   ]),
  //   height: PropTypes.oneOfType([
  //     PropTypes.string.isRequired,
  //     PropTypes.number.isRequired,
  //   ]),
  //   viewBox: PropTypes.string,
  //   children: PropTypes.node.isRequired
  // };

  // static defaultProps = {
  //   width: 800,
  //   height: 600,
  //   viewBox: "-50 -50 100 100"
  // }

  return (
    <svg
      className="grid"
      width={width}
      height={height}
      viewBox={viewBox}
      version="1.1"
      xmlns="http://www.w3.org/2000/svg"
    >
      {children}
    </svg>
  )
}
export default HexGrid
