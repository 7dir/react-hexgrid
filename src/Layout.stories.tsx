import React from "react";
import type { Meta, StoryObj } from "@storybook/react";
import Hexagon from "./Hexagon/Hexagon";

import { HexGrid, Layout, GridGenerator } from "./"

import "../.storybook/global.css";

const meta = {
    title: "Stories/Layout",
    component: Layout
} as Meta<typeof Layout>

export default meta;

type Story = StoryObj<typeof Layout>

export const Pointy: Story = {
    render: () => (
        <>
            <HexGrid width="100%">
                <Layout flat={false}>
                    {GridGenerator.hexagon(2).map((hex, i) => (
                        <Hexagon key={i} q={hex.q} r={hex.r} s={hex.s} className="tutorial" />
                    ))}
                </Layout>
            </HexGrid>
        </>
    )
}

export const Spacing: Story = {
    render: () => (
        <>
            <HexGrid width="100%">
                <Layout spacing={1.1}>
                    {GridGenerator.hexagon(2).map((hex, i) => (
                        <Hexagon key={i} q={hex.q} r={hex.r} s={hex.s} className="tutorial" />
                    ))}
                </Layout>
            </HexGrid >
        </>
    )
}