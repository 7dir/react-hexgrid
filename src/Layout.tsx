import * as React from "react"
import { Orientation } from "./models/Orientation"
import { Point } from "./models/Point"

export type Size = { x: number; y: number }

export type LayoutDimension = {
  size: Size
  orientation: Orientation
  origin: Size
  spacing: number
}
export type LayoutContextProps = {
  layout: LayoutDimension
  points: string
}

const LAYOUT_FLAT = new Orientation(
  3.0 / 2.0,
  0.0,
  Math.sqrt(3.0) / 2.0,
  Math.sqrt(3.0),
  2.0 / 3.0,
  0.0,
  -1.0 / 3.0,
  Math.sqrt(3.0) / 3.0,
  0.0,
)
const LAYOUT_POINTY = new Orientation(
  Math.sqrt(3.0),
  Math.sqrt(3.0) / 2.0,
  0.0,
  3.0 / 2.0,
  Math.sqrt(3.0) / 3.0,
  -1.0 / 3.0,
  0.0,
  2.0 / 3.0,
  0.5,
)
const defaultSize = new Point(10, 10)
const defaultOrigin = new Point(0, 0)
const defaultSpacing = 1.0

const Context = React.createContext<LayoutContextProps>({
  layout: {
    size: defaultSize,
    orientation: LAYOUT_FLAT,
    origin: defaultOrigin,
    spacing: defaultSpacing,
  },
  points: "",
})

export function useLayoutContext() {
  const ctx = React.useContext(Context)
  return ctx
}

function getPointOffset(corner, orientation: Orientation, size: Size) {
  let angle = (2.0 * Math.PI * (corner + orientation.startAngle)) / 6
  return new Point(size.x * Math.cos(angle), size.y * Math.sin(angle))
}

// TODO Refactor
function calculateCoordinates(orientation: Orientation, size: Size) {
  const corners = []
  const center = new Point(0, 0)

  Array.from(new Array(6), (x, i) => {
    const offset = getPointOffset(i, orientation, size)
    const point = new Point(center.x + offset.x, center.y + offset.y)
    corners.push(point)
  })

  return corners
}

export type LayoutProps = {
  children:
    | React.ReactElement
    | React.ReactElement[]
    | JSX.Element
    | JSX.Element[]
  className?: string
  flat?: boolean
  origin?: any
  size?: Size
  spacing?: number
}

export function Layout({
  size = defaultSize,
  flat = true,
  spacing = defaultSpacing,
  origin = defaultOrigin,
  children,
  className,
  ...rest
}: LayoutProps) {
  const orientation = flat ? LAYOUT_FLAT : LAYOUT_POINTY
  const cornerCoords = calculateCoordinates(orientation, size)
  const points = cornerCoords.map((point) => `${point.x},${point.y}`).join(" ")
  const childLayout = Object.assign({}, rest, {
    orientation,
    size,
    origin,
    spacing,
  })

  // console.log({ childLayout })

  return (
    <Context.Provider
      value={{
        layout: childLayout,
        points,
      }}
    >
      <g className={className}>{children}</g>
    </Context.Provider>
  )
}

export default Layout
