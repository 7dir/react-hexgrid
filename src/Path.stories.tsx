import React from "react";
import type { Meta, StoryObj } from "@storybook/react";
import Hexagon from "./Hexagon/Hexagon";
import { HexGrid, Layout, Path, GridGenerator, Hex } from "./";
import type { PathProps } from "./Path";


const meta = {
    title: "Stories/Path",
    component: Path
} as Meta<typeof Path>

export default meta;

type Story = StoryObj<typeof Path>

const args: PathProps = {
    start: new Hex(5, -3, -2),
    end: new Hex(-5, 1, 4),
    strokeWidth: "0.1em",
    strokeLinecap: "round",
    strokeLinejoin: "round",
    stroke: "black",
}

export const Default: Story = {
    render: () => (
        <>
            <HexGrid width="100%" height="100%">
                <Layout flat={false} spacing={1.1}>
                    {GridGenerator.hexagon(10).map((hex, i) => (
                        <Hexagon key={i} q={hex.q} r={hex.r} s={hex.s} cellStyle={{ fill: "red" }} />
                    ))}
                    <Path
                        fill="none"
                        {...args}
                    />
                </Layout>
            </HexGrid>
        </>
    )
}

export const FillTrue: Story = {
    render: () => (
        <>
            <HexGrid width="100%" height="100%">
                <Layout flat={false} spacing={1.1}>
                    {GridGenerator.hexagon(10).map((hex, i) => (
                        <Hexagon key={i} q={hex.q} r={hex.r} s={hex.s} cellStyle={{ fill: "red" }} />
                    ))}
                    <Path
                        fill="blue"
                        {...args}
                    />
                </Layout>
            </HexGrid>
        </>
    )
}