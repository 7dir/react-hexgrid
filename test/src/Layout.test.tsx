import React from "react"
import { render } from "@testing-library/react"

import { Layout } from "../../src/Layout"

test("Layout should render correctly with default props", () => {
  const { container } = render(
    <svg>
      <Layout className={"test1"}>
        <div>child</div>
      </Layout>
    </svg>,
  )
  expect(container.firstChild).toMatchSnapshot()
})

test("Layout should render correctly with custom props", () => {
  const { container } = render(
    <svg>
      <Layout
        className={"test2"}
        flat={false}
        origin={{ x: 2, y: 4 }}
        size={{ x: 12, y: 14 }}
        spacing={2.0}
      >
        <div>child</div>
      </Layout>
    </svg>,
  )
  expect(container.firstChild).toMatchSnapshot()
})
