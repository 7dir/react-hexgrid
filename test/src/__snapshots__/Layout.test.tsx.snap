// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout should render correctly with custom props 1`] = `
<svg>
  <g
    class="test2"
  >
    <div>
      child
    </div>
  </g>
</svg>
`;

exports[`Layout should render correctly with default props 1`] = `
<svg>
  <g
    class="test1"
  >
    <div>
      child
    </div>
  </g>
</svg>
`;
